### PR TITLE
Use `RetryFailedTrialCallback` in `pytorch_checkpoint` example

### LINKED ISSUE #37: Rewrite `pytorch_checkpoint` example using `RetryFailedTrialCallback`
## Motivation
Optuna v2.8.0 introduced `RetryFailedTrialCallback`, which will simplify the [pytorch_checkpoint](https://github.com/optuna/optuna-examples/blob/73cc09109c71a7e4b6d36ac33704c2eb7170e631/pytorch/pytorch_checkpoint.py) example.

## Description
The callback `restart_from_checkpoint` in the pytorch_checkpoint example can be replaced with `RetryFailedTrialCallback`. We need to change the path to the model to use the number of the first failed trial instead of the current trial's number.

### PR DESCRIPTION
**Pull request by xadrianzetx**
## Motivation

Resolves #37 

## Description of the changes

* Use `RetryFailedTrialCallback` in `pytorch_checkpoint.py` example
* Fix bug where attempt to load a checkpoint could result in `FileNotFoundError`

### CODE REVIEW

**Comment by hvy** [comment]
@not522 could you review this PR?

File: pytorch/pytorch_checkpoint.py [hunk 1]
```diff
@@ -84,9 +84,10 @@ def objective(trial):
...
     optimizer = getattr(optim, optimizer_name)(model.parameters(), lr=lr)
 
-    if "checkpoint_path" in trial.user_attrs:
-        checkpoint_path = trial.user_attrs["checkpoint_path"]
-        trial_number = trial.user_attrs["trial_number"]
+    if "failed_trial" in trial.system_attrs:
+        trial_number = trial.system_attrs["failed_trial"]
+        trial_checkpoint_dir = os.path.join(CHECKPOINT_DIR, str(trial_number))
+        checkpoint_path = os.path.join(trial_checkpoint_dir, "model.pt")
         checkpoint = torch.load(checkpoint_path)
```
Comment (not522) [inline pytorch/pytorch_checkpoint.py:91]: Could you add checking the existence of the checkpoint? If the previous trial failed before the first model saving, this line raises `FileNotFoundError`.

File: pytorch/pytorch_checkpoint.py [hunk 2]
```diff
@@ -84,9 +84,10 @@ def objective(trial):
     lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
     optimizer = getattr(optim, optimizer_name)(model.parameters(), lr=lr)
 
-    if "checkpoint_path" in trial.user_attrs:
-        checkpoint_path = trial.user_attrs["checkpoint_path"]
-        trial_number = trial.user_attrs["trial_number"]
+    if "failed_trial" in trial.system_attrs:
```
Comment (not522) [inline pytorch/pytorch_checkpoint.py:87]: Could you use `RetryFailedTrialCallback.retried_trial_number` instead?

**Review by not522 — CHANGES_REQUESTED** [review]
Thank you for your contribution! I have some comments. Could you take a look?

File: pytorch/pytorch_checkpoint.py (hunk 1 above)
Comment (xadrianzetx) [reply pytorch/pytorch_checkpoint.py:91]: Added to scope, thanks.

**Review by not522 — APPROVED** [review]
Thank you! LGTM!

**Pull request merged**